File: frontend/src/types/equipment15.ts
/** Domain types for equipment / EquipmentRecord15 */
export type EquipmentRecord15Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface EquipmentRecord15 {
  id: string;
  externalCode: string;
  displayName: string;
  status: EquipmentRecord15Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface EquipmentRecord15CreateRequest {
  externalCode: string;
  displayName: string;
  status: EquipmentRecord15Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface EquipmentRecord15ListResponse {
  items: EquipmentRecord15[];
  total?: number;
}

export interface EquipmentRecord15StatsResponse {
  activeCount: number;
}

export function isEquipmentRecord15Active(rec: EquipmentRecord15): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatEquipmentRecord15Label(rec: EquipmentRecord15): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareEquipmentRecord15ByPriority(a: EquipmentRecord15, b: EquipmentRecord15): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}

File: backend/internal/service/equipment_15.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// EquipmentRecord15Service encapsulates business rules for EquipmentRecord15.
type EquipmentRecord15Service struct {
	repo *repository.EquipmentRecord15Repository
}

// NewEquipmentRecord15Service wires the service.
func NewEquipmentRecord15Service(repo *repository.EquipmentRecord15Repository) *EquipmentRecord15Service {
	return &EquipmentRecord15Service{repo: repo}
}

// Create validates and persists a new EquipmentRecord15.
func (s *EquipmentRecord15Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.EquipmentRecord15, error) {
	now := time.Now().UTC()
	rec := &models.EquipmentRecord15{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a EquipmentRecord15 by id.
func (s *EquipmentRecord15Service) Get(ctx context.Context, id string) (*models.EquipmentRecord15, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped EquipmentRecord15 pages.
func (s *EquipmentRecord15Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.EquipmentRecord15, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing EquipmentRecord15.
func (s *EquipmentRecord15Service) Update(ctx context.Context, actor string, rec *models.EquipmentRecord15) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.UpdatedBy = actor
	rec.UpdatedAt = time.Now().UTC()
	// repository Update applies optimistic locking and increments version
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a EquipmentRecord15.
func (s *EquipmentRecord15Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds EquipmentRecord15 by external code prefix.
func (s *EquipmentRecord15Service) Search(ctx context.Context, facilityID, prefix string) ([]models.EquipmentRecord15, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *EquipmentRecord15Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}
